src/kvstore/kvstore_local.h

```diff
@@ -256,14 +260,14 @@ class KVStoreLocal : public KVStore {
   /**
    * \brief group values on keys for pull
    */
-  void GroupKVPairsPull(const std::vector<int>& keys,
-                        const std::vector<NDArray*>& values,
-                        std::vector<int> *uniq_keys,
-                        std::vector<std::vector<NDArray*>> *grouped_vals) {
+  virtual void GroupKVPairsPull(const std::vector<int>& keys,
```

Comment:
Why is `virtual` added here?

Reply:
Because the `validator` may be different in inheriting classes. That is the case for NCCL kvstore - it inherits from local kvstore to remove copy-paste code, but I can't support sparse types there.